File: pipeline/reimpl/cbtrn02c.py
"""
Reimplementation of CBTRN02C — CardDemo Daily Transaction Posting.

Function: Read daily transactions sequentially, validate each one, then either:
  - POST: write to TRANSACT-FILE, update TCATBAL, update ACCOUNT
  - REJECT: write to DALYREJS with failure reason code

Validation rules:
  100: Invalid card number (not found in XREF)
  101: Account record not found
  102: Over credit limit
  103: Transaction received after account expiration date
"""

from __future__ import annotations
from dataclasses import dataclass, field
from decimal import Decimal
from typing import Callable

from .carddemo_data import (
    AccountRecord, CardXrefRecord, TranCatBalRecord, TranRecord,
    get_db2_timestamp,
)
from .cbtrn01c import DalyTranRecord


# ── Reject record ─────────────────────────────────────────────────────────────

@dataclass
class RejectRecord:
    tran_data: DalyTranRecord
    fail_reason: int
    fail_reason_desc: str


# ── Repository interfaces ──────────────────────────────────────────────────────

class XrefRepository:
    def __init__(self, xrefs: dict[str, CardXrefRecord]):
        self._by_card = xrefs

    def find_by_card(self, card_num: str) -> CardXrefRecord | None:
        return self._by_card.get(card_num)


class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> AccountRecord | None:
        return self._accounts.get(acct_id)

    def update(self, acct: AccountRecord) -> None:
        self._accounts[acct.acct_id] = acct


class TcatbalRepository:
    """Transaction category balance store — keyed by (acct_id, type_cd, cat_cd)."""

    def __init__(self, records: dict[tuple, TranCatBalRecord] = None):
        self._records = records or {}

    def find(self, acct_id: int, type_cd: str, cat_cd: int) -> TranCatBalRecord | None:
        return self._records.get((acct_id, type_cd, cat_cd))

    def upsert(self, rec: TranCatBalRecord) -> None:
        key = (rec.trancat_acct_id, rec.trancat_type_cd, rec.trancat_cd)
        self._records[key] = rec


# ── Validation codes ──────────────────────────────────────────────────────────

FAIL_INVALID_CARD   = 100
FAIL_ACCOUNT_NF     = 101
FAIL_OVERLIMIT      = 102
FAIL_EXPIRED        = 103


# ── Result ─────────────────────────────────────────────────────────────────────

@dataclass
class PostResult:
    transaction_count: int = 0
    reject_count: int = 0
    posted_transactions: list[TranRecord] = field(default_factory=list)
    rejected_records: list[RejectRecord] = field(default_factory=list)
    return_code: int = 0
    display_lines: list[str] = field(default_factory=list)


# ── Core processing ───────────────────────────────────────────────────────────

def post_daily_transactions(
    transactions: list[DalyTranRecord],
    xref_repo: XrefRepository,
    account_repo: AccountRepository,
    tcatbal_repo: TcatbalRepository,
    logger: Callable[[str], None] = print,
) -> PostResult:
    """Post daily transactions — mirrors CBTRN02C PROCEDURE DIVISION."""
    result = PostResult()
    logger("START OF EXECUTION OF PROGRAM CBTRN02C")

    for tran in transactions:
        result.transaction_count += 1

        fail_reason = 0
        fail_desc = ""
        xref = None
        acct = None

        # 1500-VALIDATE-TRAN → 1500-A-LOOKUP-XREF
        xref = xref_repo.find_by_card(tran.tran_card_num)
        if xref is None:
            fail_reason = FAIL_INVALID_CARD
            fail_desc = "INVALID CARD NUMBER FOUND"
        else:
            # 1500-B-LOOKUP-ACCT
            acct = account_repo.find(xref.xref_acct_id)
            if acct is None:
                fail_reason = FAIL_ACCOUNT_NF
                fail_desc = "ACCOUNT RECORD NOT FOUND"
            else:
                # Check credit limit
                temp_bal = (
                    acct.acct_curr_cyc_credit
                    - acct.acct_curr_cyc_debit
                    + Decimal(str(tran.tran_amt))
                )
                if acct.acct_credit_limit < temp_bal:
                    fail_reason = FAIL_OVERLIMIT
                    fail_desc = "OVERLIMIT TRANSACTION"
                # Check expiration (only set if not already failed)
                if fail_reason == 0:
                    tran_date = tran.tran_orig_ts[:10] if tran.tran_orig_ts else ""
                    if acct.acct_expiration_date < tran_date:
                        fail_reason = FAIL_EXPIRED
                        fail_desc = "TRANSACTION RECEIVED AFTER ACCT EXPIRATION"

        if fail_reason == 0 and xref is not None and acct is not None:
            # 2000-POST-TRANSACTION
            posted = _post_transaction(tran, xref, acct, tcatbal_repo, account_repo)
            result.posted_transactions.append(posted)
        else:
            # 2500-WRITE-REJECT-REC
            result.reject_count += 1
            result.rejected_records.append(
                RejectRecord(tran, fail_reason, fail_desc)
            )

    logger(f"TRANSACTIONS PROCESSED :{result.transaction_count:09d}")
    logger(f"TRANSACTIONS REJECTED  :{result.reject_count:09d}")

    if result.reject_count > 0:
        result.return_code = 4

    logger("END OF EXECUTION OF PROGRAM CBTRN02C")
    return result


def _format_bal(d: Decimal) -> str:
    return f"{d.quantize(Decimal('0.01')):.2f}"


# ── Differential harness runner adapter ────────────────────────────────────
#
# Hardcoded scenarios dispatched by the SCENARIO input key. Each scenario
# builds its own repos + transaction list so parity tests are fully
# deterministic (no datetime, no accumulated state).
#
# Inputs:
#   SCENARIO — one of "HAPPY_GOLDEN_VECTOR", "INVALID_CARD", "ACCT_NOT_FOUND",
#              "OVERLIMIT", "EXPIRED", "MIXED_BATCH"
#
# Outputs:
#   TRANSACTION_COUNT — total transactions processed
#   REJECT_COUNT      — number rejected
#   POSTED_COUNT      — number successfully posted
#   RETURN_CODE       — 4 if any rejected, 0 otherwise
#   FAIL_CODES        — comma-separated fail codes, one per transaction
#                       ("0" means posted, "100"-"103" are reject reasons)
#   FINAL_BAL_<acct>  — final curr_bal for each account touched (for success path)


def _scenario_happy_golden_vector():
    """The W1 CobolDecimal golden vector embedded in a full CBTRN02C run.

    acct.curr_cyc_credit = 5000.00, curr_cyc_debit = 3500.75, tran_amt = 150.25
    → temp_bal = 5000.00 - 3500.75 + 150.25 = 1649.50
    credit_limit (10000.00) < 1649.50  →  False  →  NOT overlimit, transaction posts.

    After posting:
      curr_bal 1000.00 + 150.25 = 1150.25
      cyc_credit 5000.00 + 150.25 = 5150.25 (amt > 0 → credit path)
    """
    xrefs = {"4111000000001111": CardXrefRecord(
        xref_card_num="4111000000001111", xref_cust_id=1, xref_acct_id=100000001
    )}
    accounts = {100000001: AccountRecord(
        acct_id=100000001,
        acct_active_status="Y",
        acct_curr_bal=Decimal("1000.00"),
        acct_credit_limit=Decimal("10000.00"),
        acct_curr_cyc_credit=Decimal("5000.00"),
        acct_curr_cyc_debit=Decimal("3500.75"),
        acct_expiration_date="2099-12-31",
    )}
    trans = [DalyTranRecord(
        tran_id="TRN0000000000001",
        tran_type_cd="01",
        tran_cat_cd=1,
        tran_source="POS",
        tran_desc="Golden vector test",
        tran_amt=float("150.25"),
        tran_card_num="4111000000001111",
        tran_orig_ts="2026-04-09-12.00.00.000000",
    )]
    return xrefs, accounts, {}, trans


def _scenario_invalid_card():
    xrefs = {"4111000000001111": CardXrefRecord(
        xref_card_num="4111000000001111", xref_cust_id=1, xref_acct_id=100000001
    )}
    accounts = {100000001: AccountRecord(
        acct_id=100000001,
        acct_curr_bal=Decimal("1000.00"),
        acct_credit_limit=Decimal("10000.00"),
        acct_curr_cyc_credit=Decimal("0.00"),
        acct_curr_cyc_debit=Decimal("0.00"),
        acct_expiration_date="2099-12-31",
    )}
    trans = [DalyTranRecord(
        tran_id="TRN0000000000002",
        tran_type_cd="01",
        tran_cat_cd=1,
        tran_source="POS",
        tran_amt=50.00,
        tran_card_num="9999999999999999",  # not in xref
        tran_orig_ts="2026-04-09-12.00.00.000000",
    )]
    return xrefs, accounts, {}, trans


def _scenario_acct_not_found():
    """Dangling xref: card maps to an account that doesn't exist."""
    xrefs = {"4111000000002222": CardXrefRecord(
        xref_card_num="4111000000002222", xref_cust_id=2, xref_acct_id=100000002
    )}
    accounts = {}  # 100000002 missing
    trans = [DalyTranRecord(
        tran_id="TRN0000000000003",
        tran_type_cd="01",
        tran_cat_cd=1,
        tran_source="POS",
        tran_amt=25.00,
        tran_card_num="4111000000002222",
        tran_orig_ts="2026-04-09-12.00.00.000000",
    )]
    return xrefs, accounts, {}, trans


def _scenario_overlimit():
    """Credit limit 100.00, temp_bal after txn would be 250.00 → reject 102."""
    xrefs = {"4111000000003333": CardXrefRecord(
        xref_card_num="4111000000003333", xref_cust_id=3, xref_acct_id=100000003
    )}
    accounts = {100000003: AccountRecord(
        acct_id=100000003,
        acct_curr_bal=Decimal("50.00"),
        acct_credit_limit=Decimal("100.00"),
        acct_curr_cyc_credit=Decimal("50.00"),
        acct_curr_cyc_debit=Decimal("0.00"),
        acct_expiration_date="2099-12-31",
    )}
    trans = [DalyTranRecord(
        tran_id="TRN0000000000004",
        tran_type_cd="01",
        tran_cat_cd=1,
        tran_source="POS",
        tran_amt=200.00,  # 50 - 0 + 200 = 250, exceeds 100 limit
        tran_card_num="4111000000003333",
        tran_orig_ts="2026-04-09-12.00.00.000000",
    )]
    return xrefs, accounts, {}, trans


def _scenario_expired():
    """Transaction date after account expiration → reject 103."""
    xrefs = {"4111000000004444": CardXrefRecord(
        xref_card_num="4111000000004444", xref_cust_id=4, xref_acct_id=100000004
    )}
    accounts = {100000004: AccountRecord(
        acct_id=100000004,
        acct_curr_bal=Decimal("0.00"),
        acct_credit_limit=Decimal("10000.00"),
        acct_curr_cyc_credit=Decimal("0.00"),
        acct_curr_cyc_debit=Decimal("0.00"),
        acct_expiration_date="2020-01-01",  # long expired
    )}
    trans = [DalyTranRecord(
        tran_id="TRN0000000000005",
        tran_type_cd="01",
        tran_cat_cd=1,
        tran_source="POS",
        tran_amt=10.00,
        tran_card_num="4111000000004444",
        tran_orig_ts="2026-04-09-12.00.00.000000",
    )]
    return xrefs, accounts, {}, trans


def _scenario_mixed_batch():
    """Three transactions: one posts, one invalid card, one overlimit."""
    xrefs = {
        "4111000000001111": CardXrefRecord(xref_card_num="4111000000001111", xref_cust_id=1, xref_acct_id=100000001),
        "4111000000003333": CardXrefRecord(xref_card_num="4111000000003333", xref_cust_id=3, xref_acct_id=100000003),
    }
    accounts = {
        100000001: AccountRecord(
            acct_id=100000001,
            acct_curr_bal=Decimal("1000.00"),
            acct_credit_limit=Decimal("10000.00"),
            acct_curr_cyc_credit=Decimal("0.00"),
            acct_curr_cyc_debit=Decimal("0.00"),
            acct_expiration_date="2099-12-31",
        ),
        100000003: AccountRecord(
            acct_id=100000003,
            acct_curr_bal=Decimal("50.00"),
            acct_credit_limit=Decimal("100.00"),
            acct_curr_cyc_credit=Decimal("50.00"),
            acct_curr_cyc_debit=Decimal("0.00"),
            acct_expiration_date="2099-12-31",
        ),
    }
    trans = [
        DalyTranRecord(
            tran_id="TRN0000000000006",
            tran_type_cd="01", tran_cat_cd=1,
            tran_amt=75.00,
            tran_card_num="4111000000001111",
            tran_orig_ts="2026-04-09-12.00.00.000000",
        ),
        DalyTranRecord(
            tran_id="TRN0000000000007",
            tran_type_cd="01", tran_cat_cd=1,
            tran_amt=40.00,
            tran_card_num="9999999999999999",  # invalid
            tran_orig_ts="2026-04-09-12.00.00.000000",
        ),
        DalyTranRecord(
            tran_id="TRN0000000000008",
            tran_type_cd="01", tran_cat_cd=1,
            tran_amt=500.00,  # overlimit on acct 3
            tran_card_num="4111000000003333",
            tran_orig_ts="2026-04-09-12.00.00.000000",
        ),
    ]
    return xrefs, accounts, {}, trans


_CBTRN02C_SCENARIOS = {
    "HAPPY_GOLDEN_VECTOR": _scenario_happy_golden_vector,
    "INVALID_CARD": _scenario_invalid_card,
    "ACCT_NOT_FOUND": _scenario_acct_not_found,
    "OVERLIMIT": _scenario_overlimit,
    "EXPIRED": _scenario_expired,
    "MIXED_BATCH": _scenario_mixed_batch,
}


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness."""
    scenario_name = str(inputs.get("SCENARIO", "")).upper()
    if scenario_name not in _CBTRN02C_SCENARIOS:
        return {"error": f"unknown scenario: {scenario_name!r}"}

    xrefs, accounts, _tcatbal, trans = _CBTRN02C_SCENARIOS[scenario_name]()
    xref_repo = XrefRepository(xrefs)
    account_repo = AccountRepository(dict(accounts))
    tcatbal_repo = TcatbalRepository({})

    # Snapshot the original account ids so we can report final balances
    # even when the original accounts dict is mutated in place.
    acct_ids = sorted(accounts.keys())

    result = post_daily_transactions(
        trans, xref_repo, account_repo, tcatbal_repo, logger=lambda _: None,
    )

    # Build per-transaction fail code list
    fail_codes = []
    posted_idx = 0
    rejected_idx = 0
    for _ in trans:
        # We need to figure out for each tran whether it was posted or rejected.
        # post_daily_transactions preserves input order, so we walk both lists.
        # Simpler: compare the total posted + rejected to input count; for
        # ordering, rebuild by matching tran_id.
        pass
    # Walk the posted and rejected collections in order — they were appended
    # in the same order as the input transactions
    posted_ids = [t.tran_id for t in result.posted_transactions]
    rejected_ids = [r.tran_data.tran_id for r in result.rejected_records]
    rejected_by_id = {r.tran_data.tran_id: r.fail_reason for r in result.rejected_records}
    for t in trans:
        if t.tran_id in posted_ids:
            fail_codes.append("0")
        else:
            fail_codes.append(str(rejected_by_id.get(t.tran_id, 0)))

    out = {
        "TRANSACTION_COUNT": str(result.transaction_count),
        "REJECT_COUNT": str(result.reject_count),
        "POSTED_COUNT": str(len(result.posted_transactions)),
        "RETURN_CODE": str(result.return_code),
        "FAIL_CODES": ",".join(fail_codes),
    }
    for aid in acct_ids:
        out[f"FINAL_BAL_{aid}"] = _format_bal(accounts[aid].acct_curr_bal)

    return out


def _post_transaction(
    tran: DalyTranRecord,
    xref: CardXrefRecord,
    acct: AccountRecord,
    tcatbal_repo: TcatbalRepository,
    account_repo: AccountRepository,
) -> TranRecord:
    """2000-POST-TRANSACTION: update tcatbal, update account, write transaction."""
    amt = Decimal(str(tran.tran_amt))
    ts = get_db2_timestamp()

    # 2700-UPDATE-TCATBAL
    cat_key = (xref.xref_acct_id, tran.tran_type_cd, tran.tran_cat_cd)
    existing = tcatbal_repo.find(xref.xref_acct_id, tran.tran_type_cd, tran.tran_cat_cd)
    if existing is None:
        new_cat = TranCatBalRecord(
            trancat_acct_id=xref.xref_acct_id,
            trancat_type_cd=tran.tran_type_cd,
            trancat_cd=tran.tran_cat_cd,
            tran_cat_bal=amt,
        )
        tcatbal_repo.upsert(new_cat)
    else:
        existing.tran_cat_bal += amt
        tcatbal_repo.upsert(existing)

    # 2800-UPDATE-ACCOUNT-REC
    acct.acct_curr_bal += amt
    if amt >= 0:
        acct.acct_curr_cyc_credit += amt
    else:
        acct.acct_curr_cyc_debit += amt
    account_repo.update(acct)

    # 2900-WRITE-TRANSACTION-FILE
    return TranRecord(
        tran_id=tran.tran_id,
        tran_type_cd=tran.tran_type_cd,
        tran_cat_cd=tran.tran_cat_cd,
        tran_source=tran.tran_source,
        tran_desc=tran.tran_desc,
        tran_amt=amt,
        tran_merchant_id=tran.tran_merchant_id,
        tran_merchant_name=tran.tran_merchant_name,
        tran_merchant_city=tran.tran_merchant_city,
        tran_merchant_zip=tran.tran_merchant_zip,
        tran_card_num=tran.tran_card_num,
        tran_orig_ts=tran.tran_orig_ts,
        tran_proc_ts=ts,
    )

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbtrn02c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Java reimplementation of CardDemo batch program CBTRN02C — Daily Transaction Posting.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbtrn02c.py} byte-for-byte. Python is the
 * source of truth.
 *
 * <p><b>Why this program matters:</b> the credit-limit check uses the EXACT
 * calculation that the W1 CobolDecimal golden vector test was derived from:
 * <pre>
 *   temp_bal = acct_curr_cyc_credit - acct_curr_cyc_debit + tran_amt
 * </pre>
 * The CardDemo COBOL source at {@code CBTRN02C.cbl:403} computes this via a
 * COMPUTE statement with PIC S9(09)V99 operands. The W1 unit test verified
 * that {@code CobolDecimal} produces {@code 1649.50} for
 * {@code 5000.00 - 3500.75 + 150.25}; this program exercises the same math
 * in a full posting flow via the differential harness. The HAPPY_GOLDEN_VECTOR
 * scenario deliberately uses those exact values.
 *
 * <p>For parity with the Python reimpl, this class uses plain {@link BigDecimal}
 * (not {@link com.modernization.masquerade.cobol.CobolDecimal}) — the Python
 * uses plain {@code Decimal} for the arithmetic, so we match. Addition,
 * subtraction, and comparison produce byte-identical results between the two
 * for in-range values.
 *
 * <p>Validation fail codes:
 * <ul>
 *   <li>100 — card number not in XREF</li>
 *   <li>101 — card maps to a missing account (dangling xref)</li>
 *   <li>102 — over credit limit</li>
 *   <li>103 — transaction received after account expiration date</li>
 * </ul>
 */
public class Cbtrn02c implements ProgramRunner {

    private static final int FAIL_INVALID_CARD = 100;
    private static final int FAIL_ACCOUNT_NF   = 101;
    private static final int FAIL_OVERLIMIT    = 102;
    private static final int FAIL_EXPIRED      = 103;

    private static final BigDecimal BD_0_00 = new BigDecimal("0.00");

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "").toUpperCase();
        Scenario scn = buildScenario(scenarioName);
        if (scn == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        // Snapshot account ids so we can report FINAL_BAL_<id> keys in a
        // stable order even though the accounts map is mutated in place.
        List<Integer> acctIds = new ArrayList<>(scn.accounts.keySet());
        acctIds.sort(Integer::compare);

        PostResult result = postDailyTransactions(
                scn.transactions, scn.xrefs, scn.accounts
        );

        // Build FAIL_CODES list in input order (0 = posted, else reject reason)
        List<String> failCodes = new ArrayList<>();
        Map<String, Integer> rejectedById = new HashMap<>();
        for (RejectRecord r : result.rejected) {
            rejectedById.put(r.tran.tranId, r.failReason);
        }
        List<String> postedIds = new ArrayList<>();
        for (DalyTranRecord t : result.posted) {
            postedIds.add(t.tranId);
        }
        for (DalyTranRecord t : scn.transactions) {
            if (postedIds.contains(t.tranId)) {
                failCodes.add("0");
            } else {
                failCodes.add(String.valueOf(rejectedById.getOrDefault(t.tranId, 0)));
            }
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("TRANSACTION_COUNT", String.valueOf(result.transactionCount));
        out.put("REJECT_COUNT", String.valueOf(result.rejectCount));
        out.put("POSTED_COUNT", String.valueOf(result.posted.size()));
        out.put("RETURN_CODE", String.valueOf(result.returnCode));
        out.put("FAIL_CODES", String.join(",", failCodes));

        for (int aid : acctIds) {
            out.put("FINAL_BAL_" + aid, formatBal(scn.accounts.get(aid).acctCurrBal));
        }
        return out;
    }

    /**
     * Process daily transactions — mirrors Python's post_daily_transactions
     * line-by-line. Validation order:
     *   1. Lookup card in XREF  (fail 100 on miss)
     *   2. Lookup account       (fail 101 on miss)
     *   3. Credit limit check   (fail 102 if over)
     *   4. Expiration check     (fail 103 if late)
     * Valid transactions post to the account (curr_bal, cyc_credit/debit).
     */
    PostResult postDailyTransactions(
            List<DalyTranRecord> transactions,
            Map<String, CardXrefRecord> xrefs,
            Map<Integer, AccountRecord> accounts
    ) {
        PostResult result = new PostResult();

        for (DalyTranRecord tran : transactions) {
            result.transactionCount++;

            int failReason = 0;
            CardXrefRecord xref = xrefs.get(tran.tranCardNum);
            AccountRecord acct = null;

            if (xref == null) {
                failReason = FAIL_INVALID_CARD;
            } else {
                acct = accounts.get(xref.xrefAcctId);
                if (acct == null) {
                    failReason = FAIL_ACCOUNT_NF;
                } else {
                    // Credit limit check:
                    // temp_bal = cyc_credit - cyc_debit + tran_amt
                    BigDecimal tempBal = acct.acctCurrCycCredit
                            .subtract(acct.acctCurrCycDebit)
                            .add(tran.tranAmt);
                    if (acct.acctCreditLimit.compareTo(tempBal) < 0) {
                        failReason = FAIL_OVERLIMIT;
                    }
                    // Expiration check (only if not already failed)
                    if (failReason == 0) {
                        String tranDate = tran.tranOrigTs.length() >= 10
                                ? tran.tranOrigTs.substring(0, 10) : "";
                        // ISO-8601 strings compare correctly via lexicographic order
                        if (acct.acctExpirationDate.compareTo(tranDate) < 0) {
                            failReason = FAIL_EXPIRED;
                        }
                    }
                }
            }

            if (failReason == 0 && acct != null) {
                // Post the transaction
                BigDecimal amt = tran.tranAmt;
                acct.acctCurrBal = acct.acctCurrBal.add(amt);
                if (amt.signum() >= 0) {
                    acct.acctCurrCycCredit = acct.acctCurrCycCredit.add(amt);
                } else {
                    acct.acctCurrCycDebit = acct.acctCurrCycDebit.add(amt);
                }
                accounts.put(acct.acctId, acct);
                result.posted.add(tran);
            } else {
                result.rejectCount++;
                result.rejected.add(new RejectRecord(tran, failReason));
            }
        }

        if (result.rejectCount > 0) {
            result.returnCode = 4;
        }
        return result;
    }

    private static String formatBal(BigDecimal d) {
        return d.setScale(2, RoundingMode.HALF_UP).toPlainString();
    }

    // ── Scenario builder ─────────────────────────────────────────────────

    private Scenario buildScenario(String name) {
        switch (name) {
            case "HAPPY_GOLDEN_VECTOR": return scenarioHappyGoldenVector();
            case "INVALID_CARD":        return scenarioInvalidCard();
            case "ACCT_NOT_FOUND":      return scenarioAcctNotFound();
            case "OVERLIMIT":           return scenarioOverlimit();
            case "EXPIRED":             return scenarioExpired();
            case "MIXED_BATCH":         return scenarioMixedBatch();
            default: return null;
        }
    }

    private Scenario scenarioHappyGoldenVector() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000001111",
                new CardXrefRecord("4111000000001111", 1, 100000001));
        s.accounts.put(100000001, new AccountRecord(
                100000001, new BigDecimal("1000.00"), new BigDecimal("10000.00"),
                new BigDecimal("5000.00"), new BigDecimal("3500.75"),
                "2099-12-31"
        ));
        s.transactions.add(makeTran(
                "TRN0000000000001", "01", 1,
                new BigDecimal("150.25"), "4111000000001111"
        ));
        return s;
    }

    private Scenario scenarioInvalidCard() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000001111",
                new CardXrefRecord("4111000000001111", 1, 100000001));
        s.accounts.put(100000001, new AccountRecord(
                100000001, new BigDecimal("1000.00"), new BigDecimal("10000.00"),
                BD_0_00, BD_0_00, "2099-12-31"
        ));
        s.transactions.add(makeTran(
                "TRN0000000000002", "01", 1,
                new BigDecimal("50.00"), "9999999999999999"  // not in xref
        ));
        return s;
    }

    private Scenario scenarioAcctNotFound() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000002222",
                new CardXrefRecord("4111000000002222", 2, 100000002));
        // accounts intentionally empty → 100000002 will not be found
        s.transactions.add(makeTran(
                "TRN0000000000003", "01", 1,
                new BigDecimal("25.00"), "4111000000002222"
        ));
        return s;
    }

    private Scenario scenarioOverlimit() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000003333",
                new CardXrefRecord("4111000000003333", 3, 100000003));
        s.accounts.put(100000003, new AccountRecord(
                100000003, new BigDecimal("50.00"), new BigDecimal("100.00"),
                new BigDecimal("50.00"), BD_0_00, "2099-12-31"
        ));
        // temp_bal = 50 - 0 + 200 = 250 > 100 credit limit
        s.transactions.add(makeTran(
                "TRN0000000000004", "01", 1,
                new BigDecimal("200.00"), "4111000000003333"
        ));
        return s;
    }

    private Scenario scenarioExpired() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000004444",
                new CardXrefRecord("4111000000004444", 4, 100000004));
        s.accounts.put(100000004, new AccountRecord(
                100000004, BD_0_00, new BigDecimal("10000.00"),
                BD_0_00, BD_0_00, "2020-01-01"
        ));
        s.transactions.add(makeTran(
                "TRN0000000000005", "01", 1,
                new BigDecimal("10.00"), "4111000000004444"
        ));
        return s;
    }

    private Scenario scenarioMixedBatch() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000001111",
                new CardXrefRecord("4111000000001111", 1, 100000001));
        s.xrefs.put("4111000000003333",
                new CardXrefRecord("4111000000003333", 3, 100000003));
        s.accounts.put(100000001, new AccountRecord(
                100000001, new BigDecimal("1000.00"), new BigDecimal("10000.00"),
                BD_0_00, BD_0_00, "2099-12-31"
        ));
        s.accounts.put(100000003, new AccountRecord(
                100000003, new BigDecimal("50.00"), new BigDecimal("100.00"),
                new BigDecimal("50.00"), BD_0_00, "2099-12-31"
        ));
        s.transactions.addAll(Arrays.asList(
                makeTran("TRN0000000000006", "01", 1,
                        new BigDecimal("75.00"), "4111000000001111"),
                makeTran("TRN0000000000007", "01", 1,
                        new BigDecimal("40.00"), "9999999999999999"),
                makeTran("TRN0000000000008", "01", 1,
                        new BigDecimal("500.00"), "4111000000003333")
        ));
        return s;
    }

    private static DalyTranRecord makeTran(
            String id, String typeCd, int catCd,
            BigDecimal amt, String cardNum
    ) {
        DalyTranRecord t = new DalyTranRecord();
        t.tranId = id;
        t.tranTypeCd = typeCd;
        t.tranCatCd = catCd;
        t.tranAmt = amt;
        t.tranCardNum = cardNum;
        t.tranOrigTs = "2026-04-09-12.00.00.000000";
        return t;
    }

    // ── Inner data types ─────────────────────────────────────────────────

    private static class Scenario {
        final Map<String, CardXrefRecord> xrefs = new HashMap<>();
        final TreeMap<Integer, AccountRecord> accounts = new TreeMap<>();
        final List<DalyTranRecord> transactions = new ArrayList<>();
    }

    static class DalyTranRecord {
        String tranId = "";
        String tranTypeCd = "";
        int tranCatCd = 0;
        String tranSource = "";
        String tranDesc = "";
        BigDecimal tranAmt = BD_0_00;
        String tranCardNum = "";
        String tranOrigTs = "";
    }

    static class CardXrefRecord {
        final String xrefCardNum;
        final int xrefCustId;
        final int xrefAcctId;

        CardXrefRecord(String cardNum, int custId, int acctId) {
            this.xrefCardNum = cardNum;
            this.xrefCustId = custId;
            this.xrefAcctId = acctId;
        }
    }

    static class AccountRecord {
        final int acctId;
        BigDecimal acctCurrBal;
        final BigDecimal acctCreditLimit;
        BigDecimal acctCurrCycCredit;
        BigDecimal acctCurrCycDebit;
        final String acctExpirationDate;

        AccountRecord(
                int acctId, BigDecimal currBal, BigDecimal creditLimit,
                BigDecimal cycCredit, BigDecimal cycDebit, String expirationDate
        ) {
            this.acctId = acctId;
            this.acctCurrBal = currBal;
            this.acctCreditLimit = creditLimit;
            this.acctCurrCycCredit = cycCredit;
            this.acctCurrCycDebit = cycDebit;
            this.acctExpirationDate = expirationDate;
        }
    }

    static class RejectRecord {
        final DalyTranRecord tran;
        final int failReason;

        RejectRecord(DalyTranRecord tran, int failReason) {
            this.tran = tran;
            this.failReason = failReason;
        }
    }

    static class PostResult {
        int transactionCount = 0;
        int rejectCount = 0;
        int returnCode = 0;
        final List<DalyTranRecord> posted = new ArrayList<>();
        final List<RejectRecord> rejected = new ArrayList<>();
    }
}
